crawler/regelleistung.py

```diff
@@ -0,0 +1,297 @@
+import sqlalchemy
+
+from .config import db_uri
+
+
+log = logging.getLogger("regelleistung")
+log.setLevel(logging.INFO)
+
+DEFAULT_MAX_DAYS_PAST = 365*2
+
+TABLE_NAME_FCR_DEMANDS = 'fcr_bedarfe'
+URL_FCR_DEMANDS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/demands?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=FCR"
+TABLE_NAME_FCR_RESULTS = 'fcr_ergebnisse'
+URL_FCR_RESULTS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/resultsoverview?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=FCR"
+TABLE_NAME_FCR_ANONYM_RESULTS = 'fcr_anonyme_ergebnisse'
+URL_FCR_ANONYM_RESULTS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/anonymousresults?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=FCR"
+
+TABLE_NAME_AFRR_DEMANDS = 'afrr_bedarfe'
+URL_AFRR_DEMANDS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/demands?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=aFRR"
+TABLE_NAME_AFRR_RESULTS = 'afrr_ergebnisse'
+URL_AFRR_RESULTS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/resultsoverview?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=aFRR"
+TABLE_NAME_AFRR_ANONYM_RESULTS = 'afrr_anonyme_ergebnisse'
+URL_AFRR_ANONYM_RESULTS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/anonymousresults?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=aFRR"
+
+TABLE_NAME_MFRR_DEMANDS = 'mfrr_bedarfe'
+URL_MFRR_DEMANDS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/demands?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=mFRR"
+TABLE_NAME_MFRR_RESULTS = 'mfrr_ergebnisse'
+URL_MFRR_RESULTS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/resultsoverview?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=mFRR"
+TABLE_NAME_MFRR_ANONYM_RESULTS = 'mfrr_anonyme_ergebnisse'
+URL_MFRR_ANONYM_RESULTS = "https://www.regelleistung.net/apps/cpp-publisher/api/v1/download/tenders/anonymousresults?date={date_str}&exportFormat=xlsx&market=CAPACITY&productTypes=mFRR"
+
+
+def get_date_from_sql(engine, table_name, sql):
+    try:
+        with engine.begin() as conn:
+            df = pd.read_sql(sql, conn, parse_dates=["date_from"])
+            date = (df["date_from"][0]).date()
+            return date
+    except sqlalchemy.exc.ProgrammingError:
+        _, err_obj, _ = sys.exc_info()
+        if ('psycopg2.errors.UndefinedTable' in str(err_obj)):
+            log.info(
+                    f"There does not exist a table {table_name} yet."
+                )
+            return None
+        else:
+            log.error(e)
+    except Exception as e:
+        log.error(e)
+
+def get_latest_date_if_table_exists(engine, table_name):
+    sql = f"SELECT date_from FROM regelleistung.{table_name} ORDER BY date_from DESC LIMIT 1"
```

Comment:
as the db_uri and therefore also the engine contains the reference to the schema, this is not needed to be referenced here